src/libraries/System.Private.CoreLib/src/System/Threading/Overlapped.cs

```diff
@@ -0,0 +1,231 @@
+
+        [CLSCompliant(false)]
+        public NativeOverlapped* UnsafePack(IOCompletionCallback? iocb, object? userData)
+        {
+            if (_pNativeOverlapped != null)
+            {
+                throw new InvalidOperationException(SR.InvalidOperation_Overlapped_Pack);
+            }
+            _callback = iocb;
+            return AllocateNativeOverlapped(userData);
+        }
+
+        /*====================================================================
+        *  Unpacks an unmanaged native Overlapped struct.
+        *  Unpins the native Overlapped struct
+        ====================================================================*/
+        [CLSCompliant(false)]
+        public static Overlapped Unpack(NativeOverlapped* nativeOverlappedPtr)
+        {
+            ArgumentNullException.ThrowIfNull(nativeOverlappedPtr);
+
+            return GetOverlappedFromNative(nativeOverlappedPtr);
+        }
+
+        [CLSCompliant(false)]
+        public static void Free(NativeOverlapped* nativeOverlappedPtr)
+        {
+            ArgumentNullException.ThrowIfNull(nativeOverlappedPtr);
+
+            GetOverlappedFromNative(nativeOverlappedPtr)._pNativeOverlapped = null;
+            FreeNativeOverlapped(nativeOverlappedPtr);
+        }
+
+        private NativeOverlapped* AllocateNativeOverlapped(object? userData)
+        {
+            NativeOverlapped* pNativeOverlapped = null;
+            try
+            {
+                nuint handleCount = 1;
+
+                if (userData != null)
+                {
+                    if (userData.GetType() == typeof(object[]))
+                    {
+                        handleCount += (nuint)((object[])userData).Length;
+                    }
+                    else
+                    {
+                        handleCount++;
+                    }
+                }
+
+                pNativeOverlapped = (NativeOverlapped*)NativeMemory.Alloc(
+                    (nuint)(sizeof(NativeOverlapped) + sizeof(nuint)) + handleCount * (nuint)sizeof(GCHandle));
+
+                GCHandleCountRef(pNativeOverlapped) = 0;
+
+                pNativeOverlapped->InternalLow = default;
+                pNativeOverlapped->InternalHigh = default;
+                pNativeOverlapped->OffsetLow = _offsetLow;
+                pNativeOverlapped->OffsetHigh = _offsetHigh;
+                pNativeOverlapped->EventHandle = _eventHandle;
+
+                GCHandleRef(pNativeOverlapped, 0) = GCHandle.Alloc(this);
+                GCHandleCountRef(pNativeOverlapped)++;
+
+                if (userData != null)
+                {
+                    if (userData.GetType() == typeof(object[]))
+                    {
+                        object[] objArray = (object[])userData;
+                        for (int i = 0; i < objArray.Length; i++)
+                        {
+                            GCHandleRef(pNativeOverlapped, (nuint)(i + 1)) = GCHandle.Alloc(objArray[i], GCHandleType.Pinned);
```

Comment:
I'll be interested to see how the perf of this compares to async pinned handles.  For sure a win for reduction in complexity and reduction in allocations.  No immediate concerns, just a thought as I read through this.

Reply:
The only place I could find this being used was HttpListener (https://github.com/dotnet/runtime/blob/8ec62773f573f0915f2b6e1eea306fd4a84f4eb4/src/libraries/System.Net.HttpListener/src/System/Net/Windows/AsyncRequestContext.cs#L57) so maybe this would regress with this change?

Reply:
This microbenchmark has about the same performance before and after the change. The cost is dominated by an unmanaged memory block alloc/free and GC handle alloc/free that is done both before and after the change, and so the reduced managed allocations have not really moved the needle. 
```csharp
Overlapped ov = new Overlapped();
NativeOverlapped* nativeOverlapped = ov.Pack(null, null);
Overlapped.Unpack(nativeOverlapped);
Overlapped.Free(nativeOverlapped);
```

This microbenchmarks regressed about 20% when usedData is one array, 40% when userData is two arrays, and so on:
```csharp
Overlapped ov = new Overlapped();
NativeOverlapped* nativeOverlapped = ov.Pack(null, userData);
Overlapped.Unpack(nativeOverlapped);
Overlapped.Free(nativeOverlapped);
```

> This will likely show up in our HTTP.sys benchmarks as they do use this support.

I would be curious if you can observe the impact in end-to-end benchmarks.

Reply:
Agree. We’ll see when the change propagates. I’ll keep an eye out